lib/renderer/src/renderer/lib/sokol_gfx.h

```diff
@@ -0,0 +1,15732 @@
+        gl_shd = 0;
+    }
+    _SG_GL_CHECK_ERROR();
+    return gl_shd;
+}
+
+_SOKOL_PRIVATE sg_resource_state _sg_gl_create_shader(_sg_shader_t* shd, const sg_shader_desc* desc) {
+    SOKOL_ASSERT(shd && desc);
+    SOKOL_ASSERT(!shd->gl.prog);
+    _SG_GL_CHECK_ERROR();
+
+    _sg_shader_common_init(&shd->cmn, desc);
+
+    /* copy vertex attribute names over, these are required for GLES2, and optional for GLES3 and GL3.x */
+    for (int i = 0; i < SG_MAX_VERTEX_ATTRIBUTES; i++) {
+        _sg_strcpy(&shd->gl.attrs[i].name, desc->attrs[i].name);
+    }
+
+    GLuint gl_vs = _sg_gl_compile_shader(SG_SHADERSTAGE_VS, desc->vs.source);
+    GLuint gl_fs = _sg_gl_compile_shader(SG_SHADERSTAGE_FS, desc->fs.source);
+    if (!(gl_vs && gl_fs)) {
+        return SG_RESOURCESTATE_FAILED;
+    }
+    GLuint gl_prog = glCreateProgram();
+    glAttachShader(gl_prog, gl_vs);
+    glAttachShader(gl_prog, gl_fs);
+    glLinkProgram(gl_prog);
+    glDeleteShader(gl_vs);
+    glDeleteShader(gl_fs);
+    _SG_GL_CHECK_ERROR();
+
+    GLint link_status;
+    glGetProgramiv(gl_prog, GL_LINK_STATUS, &link_status);
+    if (!link_status) {
+        GLint log_len = 0;
+        glGetProgramiv(gl_prog, GL_INFO_LOG_LENGTH, &log_len);
+        if (log_len > 0) {
+            GLchar* log_buf = (GLchar*) SOKOL_MALLOC((size_t)log_len);
+            glGetProgramInfoLog(gl_prog, log_len, &log_len, log_buf);
+            SOKOL_LOG(log_buf);
+            SOKOL_FREE(log_buf);
+        }
+        glDeleteProgram(gl_prog);
+        return SG_RESOURCESTATE_FAILED;
+    }
+    shd->gl.prog = gl_prog;
+
+    /* resolve uniforms */
+    _SG_GL_CHECK_ERROR();
+    for (int stage_index = 0; stage_index < SG_NUM_SHADER_STAGES; stage_index++) {
+        const sg_shader_stage_desc* stage_desc = (stage_index == SG_SHADERSTAGE_VS)? &desc->vs : &desc->fs;
+        _sg_gl_shader_stage_t* gl_stage = &shd->gl.stage[stage_index];
+        for (int ub_index = 0; ub_index < shd->cmn.stage[stage_index].num_uniform_blocks; ub_index++) {
+            const sg_shader_uniform_block_desc* ub_desc = &stage_desc->uniform_blocks[ub_index];
+            SOKOL_ASSERT(ub_desc->size > 0);
+            _sg_gl_uniform_block_t* ub = &gl_stage->uniform_blocks[ub_index];
+            SOKOL_ASSERT(ub->num_uniforms == 0);
+            int cur_uniform_offset = 0;
+            for (int u_index = 0; u_index < SG_MAX_UB_MEMBERS; u_index++) {
+                const sg_shader_uniform_desc* u_desc = &ub_desc->uniforms[u_index];
+                if (u_desc->type == SG_UNIFORMTYPE_INVALID) {
+                    break;
+                }
+                _sg_gl_uniform_t* u = &ub->uniforms[u_index];
+                u->type = u_desc->type;
+                u->count = (uint16_t) u_desc->array_count;
+                u->offset = (uint16_t) cur_uniform_offset;
+                cur_uniform_offset += _sg_uniform_size(u->type, u->count);
+                if (u_desc->name) {
+                    u->gl_loc 
```

Comment:
do we have to include the sokol header in-tree? could it be a git submodule pointing upstream instead?

Reply:
Yes, it should probably come from external source, but we don't use submodules here. May be i'll just delete sokol renderers because we don't get much profit from them now.